### ISSUE TITLE
"foreach" alias poorly named, does not reproduce functionality and use cases with ForEach-Object

### ISSUE THREAD

**Issue by DrKittens** [PowerShell/PowerShell#23881]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

I am experiencing a problem with the foreach alias.
The foreach alias is poorly named and is the same as a language operator.
This leads to the alias functionality when expanded not been equivalent to the function and poor readability, likely due to shared name.

### Expected behavior

```console
R:\Temp> Get-ChildItem | Select Name

Name
----
New Bitmap Image.bmp
New Rich Text Document.rtf
New Text Document.txt

R:\Temp> $example = Get-ChildItem .\
R:\Temp> alias foreach

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Alias           foreach -> ForEach-Object

R:\Temp> ForEach-Object -InputObject $example { Write-Host $_.Name}
New Bitmap Image.bmp New Rich Text Document.rtf New Text Document.txt
R:\Temp> ForEach -InputObject $example { Write-Host $_.Name}
New Bitmap Image.bmp New Rich Text Document.rtf New Text Document.txt
R:\Temp> ForEach($item in $example){Write-Host $item.Name}
New Bitmap Image.bmp
New Rich Text Document.rtf
New Text Document.txt
R:\Temp> ForEach-Object($item in $example){Write-Host $item.Name}
New Bitmap Image.bmp
New Rich Text Document.rtf
New Text Document.txt
```

### Actual behavior

```console
R:\Temp> Get-ChildItem | Select Name

Name
----
New Bitmap Image.bmp
New Rich Text Document.rtf
New Text Document.txt

R:\Temp> $example = Get-ChildItem .\
R:\Temp> alias foreach

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Alias           foreach -> ForEach-Object

R:\Temp> ForEach-Object -InputObject $example { Write-Host $_.Name}
New Bitmap Image.bmp New Rich Text Document.rtf New Text Document.txt
R:\Temp> ForEach -InputObject $example { Write-Host $_.Name}
ParserError:
Line |
   1 |  ForEach -InputObject $example { Write-Host $_.Name}
     |          ~~~~~~~~~~~~
     | The parameter InputObject is not valid for the foreach statement.
R:\Temp> ForEach($item in $example){Write-Host $item.Name}
New Bitmap Image.bmp
New Rich Text Document.rtf
New Text Document.txt
R:\Temp> ForEach-Object($item in $example){Write-Host $item.Name}
ParserError:
Line |
   1 |  ForEach-Object($item in $example){Write-Host $item.Name}
     |                       ~~
     | Unexpected token 'in' in expression or statement.
```

### Error details

```console
ParserError:
Line |
   1 |  ForEach -InputObject $example { Write-Host $_.Name}
     |          ~~~~~~~~~~~~
     | The parameter InputObject is not valid for the foreach statement.


ParserError:
Line |
   1 |  ForEach-Object($item in $example){Write-Host $item.Name}
     |                       ~~
     | Unexpected token 'in' in expression or statement.
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.2
PSEdition                      Core
GitCommitId                    7.4.2
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

![image](https://github.com/PowerShell/PowerShell/assets/13168357/8b41bfea-86f7-4cc2-a791-e0b355ec3af0)

**Comment by rhubarb-geek-nz** [comment]
I absolutely agree, I never use the foreach alias and always use the cmdlet name ForEach-Object. As a guideline, I suggest use the language keyword `foreach` at the start of a command and only use `ForEach-Object` on the right hand side of a pipeline.
Alas it is too late to remove the alias.

**Comment by DrKittens** [comment]
Would not propose the removal of the alias. That would lead to breakage.

Would propose a functional ammendment be made so that they are correctly interchangeable as the issue is "some" cases work.

Eg in your example piping into the foreach alias actually works. 

The below is a valid case where the alias is actually interchangeable with the expansion to contrast with mine in the bug report where they cannot be interchanged seamlessly.

```ps
foreach ($a in (1,2),(3,4))
{
    $a | foreach {
        $_
    }
}
```

**Comment by rhubarb-geek-nz** [comment]
> The below is a valid case where the second usage of the alias

Technically there was only one usage of the `alias`, the first `foreach` was the [foreach statement](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_foreach?view=powershell-7.4) which is part of the scripting language itself, hence why it is like `if` or `while`.

The second appearance of the literal `foreach` was the only usage of the alias to [ForEach-Object](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/foreach-object?view=powershell-7.4) which is a cmdlet and part of [Microsoft.PowerShell.Core](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/?view=powershell-7.4).

I don't know the rules the scripting engine uses to determine if aliases should be used over keywords.

**Comment by 237dmitry** [comment]
> I am experiencing a problem with the foreach alias.

This has become so commonplace that it is almost impossible to get lost in aliases. In addition, they have the highest priority.

I have made it a rule for myself not to use aliases in scripts.

**Comment by DrKittens** [comment]
> > I am experiencing a problem with the foreach alias.
> 
> This has become so commonplace that it is almost impossible to get lost in aliases. In addition, they have the highest priority.
> 
> I have made it a rule for myself not to use aliases in scripts.

I agree regarding never using alias's in scripts as a best practice.
This issue actually arose from a "bad" test case that picked up "foreach" in the foreach(X in Y) usage as an alias.
The solution obviously was to ammend the test case to check for argument mode (foreach recieving value from pipeline) as that seems to be the scenario where foreach is used in the alias context. This is incrediblky obtuse compared to identifying alias usage for most other operands, even the single character % ? etc

The intent is to highlight that **sometimes** the alias works (argument mode) and in expression mode it does not and there is no implicit logic to detect which the user wants to use.

I was unaware of https://github.com/PowerShell/PowerShell/issues/3830 it appears this road has been traveled unsuccessfully.
Shame
